cpp/src/arrow/dataset/filter_test.cc

```diff
@@ -251,6 +251,76 @@ TEST_F(FilterTest, Basics) {
   ])");
 }
 
+TEST_F(FilterTest, Cast) {
+  ASSERT_RAISES(TypeError, ("a"_ == double(1.0)).Validate(Schema({field("a", int32())})));
+
+  AssertFilter("a"_.CastTo(float64()) == double(1.0),
+               {field("a", int32()), field("b", float64())}, R"([
+      {"a": 0, "b": -0.1, "in": 0},
+      {"a": 0, "b":  0.3, "in": 0},
+      {"a": 1, "b":  0.2, "in": 1},
+      {"a": 2, "b": -0.1, "in": 0},
+      {"a": 0, "b":  0.1, "in": 0},
+      {"a": 0, "b": null, "in": 0},
+      {"a": 1, "b":  1.0, "in": 1}
+  ])");
+
+  AssertFilter("a"_ == scalar(0.6)->CastLike("a"_),
+               {field("a", int32()), field("b", float64())}, R"([
+      {"a": 0, "b": -0.1, "in": 1},
+      {"a": 0, "b":  0.3, "in": 1},
+      {"a": 1, "b":  0.2, "in": 0},
+      {"a": 2, "b": -0.1, "in": 0},
+      {"a": 0, "b":  0.1, "in": 1},
+      {"a": 0, "b": null, "in": 1},
+      {"a": 1, "b":  1.0, "in": 0}
+  ])");
+
+  AssertFilter("a"_.CastLike("b"_) == "b"_, {field("a", int32()), field("b", float64())},
+               R"([
+      {"a": 0, "b": -0.1, "in": 0},
+      {"a": 0, "b":  0.0, "in": 1},
+      {"a": 1, "b":  1.0, "in": 1},
+      {"a": 2, "b": -0.1, "in": 0},
+      {"a": 0, "b":  0.1, "in": 0},
+      {"a": 2, "b": null, "in": null},
+      {"a": 1, "b":  1.0, "in": 1}
+  ])");
+}
+
+TEST_F(ExpressionsTest, ImplicitCast) {
+  ASSERT_OK_AND_ASSIGN(auto filter,
+                       InsertImplicitCasts("a"_ == 0.0, Schema({field("a", int32())})));
+  ASSERT_EQ(E{filter}, E{"a"_ == scalar(0.0)->CastTo(int32())});
+
+  auto ns = timestamp(TimeUnit::NANO);
+  ASSERT_OK_AND_ASSIGN(filter,
+                       InsertImplicitCasts("a"_ == "1990", Schema({field("a", ns)})));
+  ASSERT_EQ(E{filter}, E{"a"_ == scalar("1990")->CastTo(ns)});
+
+  ASSERT_OK_AND_ASSIGN(
+      filter, InsertImplicitCasts("a"_ == "1990" and "b"_ == "3",
+                                  Schema({field("a", ns), field("b", int32())})));
+  ASSERT_EQ(E{filter}, E{"a"_ == scalar("1990")->CastTo(ns) and
+                         "b"_ == scalar("3")->CastTo(int32())});
```

Comment:
The string "3" is implicitly cast to an int? That doesn't sound desirable.

Reply:
The string "3" is implicitly cast to int32 in this case because
- it is compared to field "b"
- field "b" is of type int32

The implicit cast would not be inserted if, for example, field "b" were of type utf8. I'll add a test demonstrating this

Reply:
Hmm, my concern is that implicit string to non-string comparisons can be quite fragile or misleading. Does SQL do this?

Reply:
This was the requested behavior as I understood it. @fsaintjacques @nealrichardson what do you think?

Reply:
Follow what CastKernel supports.

Reply:
I don't feel strongly about it either way, TBH

Reply:
CastKernel supports casting in this fashion, but I think @pitrou 's concern is whether this should be implicitly inserted

Reply:
Right.

Reply:
@pitrou they're not on by default and this is a stop gap to allow filters like `"ts"_ > "2000-01-01"` and `"f64"_ > 100` to just work. Query validation/binding is a nuanced problem for which we need more design discussion. In a follow up PR we will probably extract `InsertImplicitCasts` and `Expression::Validate` into a separate interface which will (among other things) allow more control over what implicit casts may be inserted.

Reply:
Fair enough.